Cloud Error Reporting Exception Logger Unit Tests

iantalarico opened the pull request.

jskeet commented on apis/Google.Devtools.AspNet/Google.Devtools.AspNet.Tests/CloudErrorReporting/CloudErrorReportingExceptionLoggerTest.cs:
```diff
@@ -0,0 +1,166 @@
+            CloudErrorReportingExceptionLogger logger = GetLogger(mockClient.Object);
+            logger.Log(CreateComplexContext());
+
+            mockClient.Verify(client => client.ReportErrorEvent(FormattedProjectId, IsComplexContext(), null), Times.Once());
```
Can't you just move `FormattedProjectid, IsComplexContext(), null` into the `Setup` call, and call `VerifyAll` or whatever it's called?

iantalarico replied:
Looks like I can :) Done.

jskeet commented on apis/Google.Devtools.AspNet/Google.Devtools.AspNet.Tests/CloudErrorReporting/CloudErrorReportingExceptionLoggerTest.cs:
```diff
@@ -0,0 +1,166 @@
+            );
+        }
+
+        private Mock<ReportErrorsServiceClient> GetClientMock()
```
I think I'd personally just inline this in each usage - it makes each test simpler to read by depending on fewer things, and we're not actually using any logic that would need duplication.

iantalarico replied:
Done.

iantalarico commented:
Thanks, PTAL.

iantalarico commented:
Hey Jon,

I'm a little lost why some of the tests are failing they are all passing locally (on windows). I tried to set up a linux machine to test them but wasn't able to get it working.  Do you have instructions on how to do this? Or do you have any idea why these are failing?

Thanks
Ian

jskeet commented:
@iantalarico: I don't have specific instructions, but you should be able to follow the .NET installation instructions at http://dot.net, then just clone the repo, restore and run the tests.

It looks like it's only the complex event aspect that fails. You might want to log `exceptionContext` from `CreateComplexContext` before returning it...

iantalarico commented:
@jskeet: I'm still having some issues getting this set up.  It looks like I'm stuck on an issue you filed awhile back.  https://github.com/dotnet/cli/issues/3073  I'm unsure what exactly to do with the script they mention.

jskeet commented:
Okay - I have a spare Linux box I can try it on; I'll give that a go tomorrow.

iantalarico commented:
If you also have some time to help me get this set up tomorrow that would also be appreciated :) I'm not sure what I am doing wrong.

I am having trouble running any tests.

jskeet commented:
Of course, this isn't a .NET Core solution, which is why that's causing problems.
Basically, look in `build.sh` for how we run tests using mono.

To run them manually on my Ubuntu 16.04 box, I changed to the `apis/Google.Devtools.AspNet/Google.Devtools.AspNet.Tests` directory and ran

```sh
mono bin/Release/net451/ubuntu.16.04-x64/dotnet-test-xunit.exe bin/Release/net451/ubuntu.16.04-x64/Google.Devtools.AspNet.Tests.dll
```

Still looking into the failure itself.

jskeet commented:
Okay, I've worked out what's broken: on Mono, we end up with a line number of 0 and an empty file path, even in the complex context. Basically, the exception stack trace doesn't have the relevant information - I'm not sure what you can do about that.

You might want to detect that you're running in an environment that doesn't give you full stack trace information, and not test for that in `IsComplexContext` if so.

iantalarico commented:
Thanks for the help Jon!  I still couldn't get this working on Debian 8.  Switched to Ubuntu 16 and it all worked.

I added a check to make sure the OS is windows which seemed to fix the test. Hopefully this will pass on the CI too.

jskeet commented:
LGTM.
